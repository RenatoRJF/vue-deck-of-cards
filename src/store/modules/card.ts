import { Module, VuexModule, Action } from "vuex-module-decorators";
import api from "../../services/api";

@Module({ namespaced: true })
export default class Card extends VuexModule {
  @Action
  async createCards({ deckId, cards }: { deckId: string; cards: string }) {
    if (!cards) return;

    const numOfCards = cards.split(",").length;

    try {
      // Draw cards into the deck
      await Promise.all(
        [...Array(numOfCards).keys()].map(() => api.get(`deck/${deckId}/draw/?cards=${cards}`))
      );
    } catch (error) {
      throw new Error(error);
    }
  }
}
