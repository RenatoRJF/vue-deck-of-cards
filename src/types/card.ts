export type Card = {
  value: string;
  suit: string;
  isValid?: boolean;
};
