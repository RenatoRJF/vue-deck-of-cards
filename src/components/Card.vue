<template>
  <div class="card">
    <div class="card__content" :style="`animation-delay: ${index / 10}s`">
      <div class="card__value-suit">
        <span class="value">{{ card.value }}</span>
        <img :src="getImageUrl(card.suit)" class="suit" alt="card symbol" />
      </div>

      <div class="card__value-suit bottom">
        <span class="value">{{ card.value }}</span>
        <img :src="getImageUrl(card.suit)" class="suit" />
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";
import { Card as CardType } from "../types/card";

@Component
export default class Card extends Vue {
  @Prop() cards!: CardType;
  @Prop() index!: number;

  private getImageUrl(img: string) {
    const images = require.context("../assets/", false, /\.png$/);

    return images("./" + img + ".png");
  }
}
</script>

<style lang="scss" scoped>
.card {
  padding: 20px;
  box-sizing: border-box;
  width: 20%;
  min-width: 230px;
  opacity: 1;
}

@keyframes animate-card {
  0% {
    opacity: 0;
  }
  99% {
    opacity: 1;
  }
}

.card__content {
  background-color: #fff;
  height: 300px;
  border-radius: 20px;
  box-shadow: 0 0 15px rgba(0, 0, 0, 0.5);
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: animate-card 0.3s ease-in;
  -o-animation: animate-card 0.3s ease-in;
  -moz-animation: animate-card 0.3s ease-in;
  -webkit-animation: animate-card 0.3s ease-in;
}

.card__value-suit {
  padding: 20px 30px;
  display: flex;
  align-items: center;

  .value {
    font-size: 35px;
    font-weight: bold;
  }

  .suit {
    transform: rotate(180deg);
    margin: 0 5px;
  }

  &.bottom {
    transform: rotate(180deg);
  }
}
</style>
