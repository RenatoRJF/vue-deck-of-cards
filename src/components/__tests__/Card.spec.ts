import { mount } from "@vue/test-utils";
import Card from "@/components/Card.vue";
import { Card as CardType } from "../../types/card";

describe("Card.vue", () => {
  let card: CardType;

  beforeEach(() => {
    card = { value: "2", suit: "H" };
  });

  test("renders without crashing", () => {
    const wrapper = mount(Card, {
      propsData: { data: card }
    });
    expect(wrapper).toMatchSnapshot();
  });

  test("renders props when passed passed", async () => {
    const wrapper = mount(Card, {
      propsData: { data: card, index: 0 }
    });
    expect(wrapper.find(".value").text()).toBe(card.value);
    expect(wrapper.find("img").attributes("data-img")).toBe(`${card.suit}.png`);
    expect(wrapper).toMatchSnapshot();
  });
});
