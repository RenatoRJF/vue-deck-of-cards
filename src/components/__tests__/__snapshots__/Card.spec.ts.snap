// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue renders props when passed passed 1`] = `
<div class="card">
  <div class="card__content" style="animation-delay: 0s;">
    <div class="card__value-suit"><span class="value">2</span> <img src="" alt="card symbol" data-img="H.png" class="suit"></div>
    <div class="card__value-suit bottom"><span class="value">2</span> <img src="" class="suit"></div>
  </div>
</div>
`;

exports[`Card.vue renders without crashing 1`] = `
<div class="card">
  <div class="card__content" style="animation-delay: NaNs;">
    <div class="card__value-suit"><span class="value">2</span> <img src="" alt="card symbol" data-img="H.png" class="suit"></div>
    <div class="card__value-suit bottom"><span class="value">2</span> <img src="" class="suit"></div>
  </div>
</div>
`;
